Create morevp.md [WIP]

smartcontracts opened the pull request:
Adding initial version of MoreVP. Still WIP.

---

#### TODO:
##### Low Priority:
[ ] Add diagrams
[ ] Rewrite proofs
[ ] Add links to code

smartcontracts committed:
Create morevp.md [WIP] [NOTREADY]

Adding initial version of MoreVP. Still WIP - not ready for review.

smartcontracts committed:
Update morevp.md

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,277 @@
+
+#### Deposit Transaction
+
+A deposit transaction is a special type of transaction that creates a new balance on the Plasma chain. This is similar to a Bitcoin "coinbase" transaction. In practice, this takes the form of a transaction with a single “special” input.
```
I would add that for a deposit transaction to occur, user must deposit funds into the Plasma contract on rootchain.

smartcontracts replied:
I leave this out because it's possible for the operator to create an invalid deposit transaction that does not correspond to a deposit.

evabeylin replied:
I would include both of those as potential deposit txns. 

This document should assume that people will read it as a guide and implement without doing 100% of the research that you've done - so i) deposit into plasma contract will ensure their understanding of how deposits interact with the rootchain and ii) operator invalid deposit will ensure they are aware of operator risks.

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,277 @@
+
+A transaction is “canonical” if none of its inputs were previously spent in any other transaction. The definition of “previously spent” depends on whether or not the transaction in question is included in the Plasma chain.
+
+If the transaction was included in the chain, an input to that transaction would be considered previously spent if another transaction also spending the input was included in the chain *before* the transaction in question. Transaction position is determined by the tuple (block number, transaction index). If the transaction was not included in the chain, an input to that transaction would be consi ...
```
would move "Transaction position is determined..." as the first sentence to this paragraph

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,277 @@
+
+#### Valid Transaction
+
+A deposit transaction is “valid” if and only if it corresponds exactly to a single deposit on the root chain and is the first deposit transaction to correspond to that deposit. A spend transaction is “valid” if and only if it is exitable, canonical, and only stems from valid transactions (i.e. all transactions in the history are also valid transactions). Note that a transaction would therefore be ...
```
when is an exitable transaction not necessarily valid? - per your definition above an exitable transaction is if transaction is correctly formed (inputs older than ouputs, signatures) and presumably those inputs/outputs are valid. if there is a circumstance I would briefly spell out the specific case it's not valid.

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,277 @@
+
+### Basic Specification
+
+The MoreVP exit protocol allows the owners of both inputs and outputs to a transaction to attempt an exit. 
```
add: "...to ensure that the owner of the input can secure their funds in the case that the transaction results in being non-canonical (e.g., other inputs are invalid, outputs are invalid) or users need to exit the Plasma chain before the transaction is processed."

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,277 @@
+
+## Exit Protocol Implementation
+
+### Motivation
```
I would move this motivation right after introduction / before definitions (after row 18) as this frames up how and why this is the implementation

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,277 @@
+
+## Exit Protocol Implementation
+
+### Motivation
```
I would split this section. 

Move "Motivation" stuff right after glossary and before basic specification (as this is introducing the relevance of the spec to the reader) (Motivation = rows 83 to 85.)

Rows 87 - 89 are not motivation but seem more like additional detail describing the basic spec you outlined, would just add this to the descriptive paragraph you already have.

smartcontracts replied:
The basic spec is supposed to give intuition for the game. The implementation is in the context of the EVM, so it requires additional motivation.

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,277 @@
+Note that this challenge means it’s possible for an honest user to lose `exit bond` as they might not be aware their transaction is non-canonical. We address this attack vector and several mitigations in detail later.
+
+
+#### Joining an Exit
```
I don't see mass exits in this PR but I would make this section Mass exits and "piggybacking" as a subset, as it's parallel to a mass exit just from the same transaction.

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,277 @@
+
+##### Mitigations
+
+###### Bond Sharing
```
move this after small exit bond as small exit bond directly relates to the mallory bob situation above

smartcontracts replied:
All of these mitigations relate to the mallory/bob situation above.

smartcontracts committed:
Saving changes to morevp.md

smartcontracts committed:
Update morevp.md

Made each sentence a line, removed extra whitespace.

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+
+#### In-flight Transaction
+
+A transaction is considered to be “in-flight” if it has been broadcast but has not yet been included in the Plasma chain.
```
are transactions that have been broadcast but the inclusion or not is unknown considered in-flight? I think they should, but not sure if it plays well with the rest of the doc.

smartcontracts replied:
I see what you're saying here and will see if I can explain that more clearly. It's definitely a case in which the MoreVP exit is necessary.

smartcontracts replied:
Added some more clarification "A transaction may be in-flight from the perspective of an individual user if that user does not have access to the block in which the transaction is included." Let me know if this is useful or unclear.

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+
+#### Valid Transaction
+
+A deposit transaction is “valid” if and only if it corresponds exactly to a single deposit on the root chain and is the first deposit transaction to correspond to that deposit.
```
not sure why this definition (valid/invalid deposit tx). How could a deposit be not valid? Maybe "deposit transaction" is an unnecessary abstraction here, since really there isn't such a thing, in strict terms. A deposit is just an output (position in the chain), recognized by the root chain contract (to exit immediately) and by the chain operator (hopefully). Thereby adjusting the term "transaction" to a deposit might be more confusing than clariyfing, because you need to consider it's validity etc.

smartcontracts replied:
Originally I wanted to distinguish between deposits and spends as deposits can legitimately create value "out of nowhere," while spend transactions cannot. I think you're right that the distinction is confusing.

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+
+The MoreVP exit protocol allows the owners of both inputs and outputs to a transaction to attempt an exit.
+
+The owner of an input `in` to a transaction `tx` must prove that:
```
I had issues with that way of putting it, but now it reads good and natural!

The interesting part is that we may need to provide a way for such owner to "atomically" prove that `tx` isn't canonical, atomically with starting an in-flight exit, which would save such owner the exit_bond. That would appear when a owner spends `in` together with `other_in`, and then sees a competing spend of `other_in` in the chain. The owner should have a possibility of not losing the "first" exit bond required.

But thats 2 steps ahead so nevermind, looks all clear now!

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+
+The above game correctly selects the inputs or outputs that are eligible to exit.
+However, we still need to enforce an ordering on exits to ensure that all outputs created by valid transactions will be paid out before any output created by an invalid transaction.
+We do this by ordering every exit by the position of the *youngest input* to the transaction referenced in each exit.
```
add "... regardless of whether its the `in` or `out` exiting".

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+3. `out` is not spent.
+
+
+#### Priority
```
include how in-flight exit related priority relates to regular exit priority (i.e. priority is calculated differently, but end up "used" in the same priority queue, i.e. on par with regular exits).

I also thing detailed interactions between the two exit protocols need being specified, namely that you cannot in-flight exit an output that is being exited regularly, probably vice versa + other more exotic and gory scenarios

smartcontracts replied:
Added a new section "Combining with Plasma MVP Exit Protocol"

pdobacz replied:
@kfichter what about the 2nd paragraph, do you think we should be more specific about the correctness of these interactions? is there a list of conditions that must be checked extra for the two protocols to not interfere badly?

kasima replied:
Also curious about unintended behavior from the two protocols interacting. Though, I would be comfortable putting it beyond the scope of this document as "needs additional research."

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+We do this by ordering every exit by the position of the *youngest input* to the transaction referenced in each exit.
+
+
+## Exit Protocol Implementation
```
I'm still reluctant towards the term "Implementation" being used here... to not be confused with code.

How about we call **this** the protocol instead, as it sounds like a protocol, and what you called "Exit protocol" above, call that something like "Exit rules" or "Exit constraints" or "Protocol desired outcome" or "Desired mechanism" (as in mechanism design and reverse game theory)...?

Protocol has an air about it that makes me think about messages being exchanged as time passes (so what follows in this section), whereas the preceding section sounds more like a static set of rules/constraints/assumptions, the "holy grail" that the protocol accomplishes, if played out.

Then you can call "the implementation" the actual `.sol` and client implementation

pdobacz replied:
@kfichter note that there are two sections called "Exit Protocol" now. So if you agree with my argumentation to not use the word "implementation" in this context, the initial section should be renamed

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+Additionally, the MoreVP exit protocol is not necessary in all cases.
+Previous work shows that we can use the Plasma MVP exit protocol without confirmation signatures for any transaction included before an invalid (or, in the case of withheld blocks, potentially invalid) transaction.
+We therefore only need to make use of the MoreVP protocol for the set transactions that are in-flight when a Plasma chain becomes byzantine.
+This also means that we can force deposit transactions to use the Plasma MVP exit protocol as these transactions would never be in-flight.
```
one of the statements you'd be able to avoid, if you not call deposits "deposit transactions", see other suggestion on that

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+
+To account for this problem, we allow exits to be ambiguous about canonicity.
+Users can start MoreVP exits with the *assumption* that the referenced transaction is canonical.
+Other owners of inputs or outputs to the transaction can then join (or “piggyback”) the exit.
```
Maybe, let's just use "piggyback" as the term for that everywhere, it's settled in. So, suggestion to never use the word "join" and rephrase as:

"owners (...) can then piggyback on the exit, indicating that they would like to exit their inputs/outputs resp. during the game, and attesting that they're entitled to it"

You also give a heads-up explanation of what such piggybacking means

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+The MoreVP exit protocol requires the use of a “challenge-response” mechanism, whereby users can submit a challenge but are subject to a response that invalidates the challenge.
+To give users enough time to respond to a challenge, the exit process is split into two “periods.” When challenges are subject to a response, we require that the challenges be submitted before the end of the first exit period and that responses be submitted before the end of the second.
+We define each period to have a length of half the minimum exit period (MEP).
+Currently, the MEP is set to 7 days, so each period has a length of 3.5 days.
```
Id define `MEP/2` as a new abbreviation (same suggestion as with limbo cancels) and state explicitly, that with this protocol in place we require, that all watchers validate the chain at least every <that new term>, not every `MEP` as previously required

pdobacz replied:
@kfichter what about noting that:

> watchers validate the chain at least every , not every MEP as previously required

Oops, sorry, now I see my comment got severed somehow. It was supposed to say "watchers validate the chain at least every min(REP, /new abbreviation/), not every min(MFP, REP) as previously required" - with some extra corrections applied

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+
+#### Starting the Exit
+
+For simplicity, we block deposit transactions from using the MoreVP exit protocol.
```
ditto, no use for this note, if deposit =/= a transaction, which I think is more natural

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+
+We provide several possible mechanisms that allow a user to prove a transaction is exitable.
+First, remember that deposit transactions are exitable by default.
+Two ways in which spend transactions can be proven exitable are as follows:
```
use symbols, i.e. `tx` is "the transaction" and `input_tx1, ... input_txn` the other transactions

pdobacz replied:
also clarify which signatures are checked to be correct - these that are included in `tx` and pointing towards `newowners'` of `input_txs` or do you go one step further back

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+Another user can challenge one of these signatures by presenting some transaction that created an input such that the true input owner did not sign the signature.
+In this case, the exit would be blocked entirely and the challenging user would receive `exit bond`.
+
+Option (1) checks that a transaction is exitable when the exit is started.
```
I think we should leave a `TODO` self note to groom this part as soon as the option is chosen

smartcontracts replied:
Would this be better as a markdown comment (hidden from view) or as a visible TODO?

pdobacz replied:
@kfichter hmmm, not sure. I'd say visible, so that the readers see where the doc is headed and that we don't forget too soon, that we've left a todo

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+
+#### Proving Canonicity
+
+Whether unspent inputs or unspent outputs are paid out in an exit depends on the canonicity of the referenced transaction.
```
add "... and is fully independent of the piggybackings that happen" or similar, let's explicitly steer clear of this old confusion

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+
+The process of determining canonicity involves a challenge-response game.
+In the first period of the exit, any user may reveal a competing transaction that potentially makes the exiting transaction non-canonical.
+This competing transaction must be exitable, must share an input with the exiting transaction, and must be included in the Plasma chain.
```
"must be included in the Plasma chain." not what we talked about I think.

How else would Alice present her `tx1`, when she sees a competing `tx2` doing an in-flight, when neither tx is seen included?

Also this is conflicting with the prior statement, that if a competitor _exists_ then neither tx is canonical

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+This competing transaction must be exitable, must share an input with the exiting transaction, and must be included in the Plasma chain.
+Multiple competing transactions can be revealed during this period, but only the oldest presented transaction is considered for the purposes of a response.
+
+If any transactions have been presented during the first period, any other user can respond to the challenge by proving that the exiting transaction is actually included in the chain before the oldest presented competing transaction.
```
I'd leave a `TODO`/placeholder for a diagram to depict the state machine of a single tx being inflight exited, in terms of determining canonicity. Somethin along the lines of `canonical -> non-canonical -> canonical` where the arrows are conditional on stuff being submited to the root chain/revealed

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+We address this attack vector and several mitigations in detail later.
+
+
+#### Joining an Exit
```
ditto, not ok with having two terms for that. Let's just use "piggyback" everywhere (or if you'd like to reconsider, let's forget piggybacking and just join everywhere, but that might be hard). Multiple occurrences of that here. Paragraphs like [this](https://github.com/omisego/research/pull/44/files#diff-056213b9987c8cd9d93d6615e132221fR213) that use both will become confusing if they aren't already

smartcontracts replied:
Replacing all occurrences of "join" with "piggyback."

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+Owners of both inputs and outputs to a transaction may want to start an exit in the case that they would receive the funds from the exit.
+However, we want to avoid the gas cost of repeatedly publishing and proving statements about the same transaction.
+We therefore allow owners of inputs or outputs to a transaction to join (“piggyback”) an existing exit that references the transaction.
+This effectively “caches” the result of the “exitable” and “canonical” checks.
```
not sure what you mean here, "caches"? You used the word "preempt" before, I think it works better

smartcontracts replied:
Idea was that canonicity is determined once for a single transaction instead of once for each exit. I need to think more about this one.

smartcontracts replied:
Just going to remove this line to avoid confusion.

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+If the chain is byzantine, not piggybacking could potentially mean loss of funds.
+
+
+### Attack Vectors and Mitigations
```
sorry to not pick this up earlier - how about moving this section after alicebobbings or as the final part of alicebobbings? The flow could be easier, if we elaborate on the cases that work well before delving into the problematic ones

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+5. Mallory spends `UTXO1` in `TX2`.
+6. In period 1 of the exit for `TX1`, Mallory challenges the canonicity of `TX1` by revealing `TX2`.
+7. No one is able to respond to the challenge in period 2, so `TX1` is determined to be non-canonical.
+8. After period 2, Mallory receives Bob’s `exit bond`.
```
in the section where you describe the canonicity game, I think it is implied that that happens after period 1 not 2

smartcontracts replied:
Updated the canonicity game with the line "If this response is given before the end of the second period..." to make clear that Mallory receives the bond after period 2.

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+6. In period 2, someone reveals `TX2` spending `UTXO2`.
+This challenger receives Mallory’s `piggyback bond`.
+7. Alice is honest, so she hasn’t spent `UTXO1` in any transaction other than `TX1`.
+8. After period 2, Mallory’s `exit bond` is refunded.
```
note explicitly that no-one exits any of the UTXOs perhaps? If yes, then that applies to all scenarios

smartcontracts replied:
Added to each scenario.

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+4. Mallory spends `UTXO1` in `TX2`.
+5. `TX2` is included in a withheld block.
+`TX1` is not included in a block.
+6. Bob starts an exit referencing `TX1` and places `exit bond`.
```
I think it is not rational for Bob to do that, he gains nothing just loses `exit_bond`. Bob should not process the incoming transaction until he sees it included in the chain.

smartcontracts replied:
I think I agree with this. In reality, the owner of an input to the transaction should typically be the one to start the exit to provide better guarantees for Bob. Need to think more.

evabeylin replied:
What if Bob wasn't allowed to exit `TX1` unless he can provide proof that the `in` of `TX1` wasn't spent either. Basically for the *Desired Exit Mechanism* (rows 90-98), add the equivalent of row 93 for the owner, so they can't exit unless they prove both the output and the inputs haven't been spent. This way if the TX2 is included, Bob wouldn't lose his bond as his exit wouldn't be valid in the first place.

smartcontracts replied:
@evabeylin I'm not sure I understand what you're saying.

evabeylin replied:
I'm questioning whether it's possible to invalidate Bob's exit from TX1 in the first place, by requiring that when he exits he must prove that the input into the txn is valid, in addition to proving the the output utxo2 is valid, (similar to the design exit mechanism you have in row 93: `in` is not spent in any transaction other than `tx`). If for an exit to be valid, the owner has to prove that the input to the exiting txn is actually spent in *that txn* then you may mitigate other users from honestly losing their bonds. (?)

jepidoptera replied:
@kfichter but Mallory owns the only input to tx1...

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+8. After period 2, the user who responded to the challenge receives Mallory’s `exit bond`.
+
+
+### Mallory attempts 
```
something's missing here

pdobacz commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,648 @@
+
+1. Alice spends `UTXO1` in `TX1` to Bob, creating `UTXO2`.
+2. `TX1` is included in (valid) block `N`.
+3. Operator creates invalid deposit transaction `TX2`, creating `UTXO3`.
```
oh, ok, so the "deposit transaction" comes from here. I think it's not necessary, you just say that "Operator spends a non-existent UTXO3 in TX3". In any case I think here Operator would need to do two invalid transactions i.e. bad deposit -> "good", included spend of that -> the exited transaction, otherwise Operator doesn't have exitability

pdobacz commented:
I think we should delve into the `exitable_at` part of the exit priority, i.e. it is rather clear that the priorities inside the exit queue are fine, still I don't know whether the delay requirement (`exitable_at`) is calculated such that it doesn't introduce some unwanted behavior.

Assuming that we've got the `exitable_at` for regular exits defined as
```
max(exit_request_block.timestamp + MFP, utxo_submission_block.timestamp + MFP + REP)
```
(where `MFP` and `REP` are some constants related to the frequency of validating the chain),

let's include the counterpart of that in in-flight exits terms here. I don't suspect there's anything dangerous going on here, but I'd rather we put such formula explicitly here, so that we are sure that one cannot somehow break the `processExit` process.

smartcontracts committed:
Update morevp.md

Updated for comments by @pdobacz

kfichter commented:
Re: `exitable_at`, the formula is exactly the same, except that `utxo_submission_block.timestamp` comes from the youngest input instead of the referenced output.

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+#### Deposit
+
+A deposit creates a new output on the Plasma chain.
+Although deposits are typically represented as transactions that spend some "special" input, we do not treat deposits as transactions for the purpose of the MoreVP exit protocol.
```
Might be worth explicitly adding "Deposits are safely covered by the standard MVP exit protocol."

Minor: move the "Deposit" section above "Spend Transaction" section to group the transactions together for flow.

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+
+#### Canonical Transaction
+
+A transaction is “canonical” if none of its inputs were previously spent in any other transaction, i.e. that the transaction is the oldest among all its competitors.
```
...oldest among all its competitors, if any.

smartcontracts replied:
Our definition of competitors includes a transaction as a competitor to itself to simplify the mathematical definition. Can change if that's too confusing.

kasima replied:
Missed that. It's fine the way it is.

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+#### Exitable Transaction
+
+If a transaction is “exitable,” then a user may attempt to start an exit that references the transaction.
+A spend transaction can be called “exitable” if the transaction is correctly formed (e.g. more input value than output value, inputs older than outputs, proper structure) and is properly signed by the owners of the transaction’s inputs.
```
Swapping this with the previous sentence would flow better.

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+
+A spend transaction is “valid” if and only if it is exitable, canonical, and only stems from valid transactions (i.e. all transactions in the history are also valid transactions).
+Note that a transaction would therefore be considered invalid if even a single invalid transaction is present in its history.
+An exitable transaction is not necessarily a valid transaction, but all valid transactions are, by definition, exitable.
```
This is a key point that could deserve some more explanation. Or mention that the exit protocol (by game and by priority) prevents successful exits of transactions that are not valid, even though they are "exitable".

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+### Desired Exit Mechanism
+
+The MoreVP exit protocol allows the owners of both inputs and outputs to a transaction to attempt an exit.
+We want to design a mechanism that allows these inputs and outputs to be withdrawn under the following conditions.
```
...allow these inputs _or outputs, but not both,_ to be withdrawn...

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+
+#### Deposit
+
+A deposit creates a new output on the Plasma chain.
```
add: ...by depositing in a Plasma contract on the rootchain and the other ways etc...

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+#### Timeline
+
+The MoreVP exit protocol requires the use of a “challenge-response” mechanism, whereby users can submit a challenge but are subject to a response that invalidates the challenge.
+To give users enough time to respond to a challenge, the exit process is split into two “periods.” When challenges are subject to a response, we require that the challenges be submitted before the end of the first exit period and that responses be submitted before the end of the second.
```
would add: 
1) why a challenge is needed in the first place (e.g., to ensure that the rootchain is getting accurate information re: our talk earlier this week)
2)  why MFP is 7 days (re: maximum time required for all txns to exit if chain is byzantine)

this is something many don't understand (e.g., Rick Dudley) and will directly impact how readers understand the UX of plasma, best to have it laid out that this is a valuable mechanism timeline

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+
+Any user may initiate an exit by presenting a spend transaction and proving that the transaction is exitable.
+The user must submit a bond, `exit bond`, for starting this action.
+This bond is later used to cover the cost for other users to publish statements about the canonicity of the transaction in question.
```
would specify that "publishing statement about canonicity" is a challenge

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+
+We still need to provide a deterministic ordering of exits by some priority.
+MoreVP exits are given a priority based on the position in the Plasma chain of the most recently included (youngest) input to that transaction.
+Unlike the MVP protocol, we give each input and output to a transaction the same priority.
```
maybe add what benefit this provides vs MVP? otherwise unclear rationale

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+
+The process of determining canonicity involves a challenge-response game.
+In the first period of the exit, any user may reveal a competing transaction that potentially makes the exiting transaction non-canonical.
+This competing transaction must be exitable and must share an input with the exiting transaction, but does not have to be included in the Plasma chain.
```
add why: that the competing txn doesnt have to be included in the chain because it ensures that if users conduct other txns with the same input during an exit's challenge period, their youngest txn will take priority..

jepidoptera replied:
I think the competing transaction doesn't have to be included in the chain because the exiting transaction might not be either.  In that case, neither is canonical.

> If the transaction was not included in the chain, an input to that transaction would be considered previously spent if another transaction also spending the input is *known to exist*.

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+2. `TX1` is included in block `N`.
+3. Mallory spends `UTXO2` in `TX2`.
+4. Mallory starts an exit referencing `TX1` and places `exit bond`.
+5. Mallory piggybacks onto the exit and places `piggyback bond`.
```
add the exit is referencing TX2 for piggyback

kasima replied:
Do you mean `UTXO2`? Mallory is trying to exit from `TX1`.

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+
+
+### Mallory attempts to exit a spent input:
+
```
I think the better description of this is she attempts to exit inputs that haven't been included yet. If at least one of them was spent and included then she wouldn't be losing both exit bonds.

evabeylin commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+Owners of both inputs and outputs to a transaction may want to start an exit in the case that they would receive the funds from the exit.
+However, we want to avoid the gas cost of repeatedly publishing and proving statements about the same transaction.
+We therefore allow owners of inputs or outputs to a transaction to piggyback an existing exit that references the transaction.
+
```
I would specify that the piggyback can be used for either 1) inputs/outputs piggybacking onto an exit or 2) the user piggybacking their own exit to withdraw funds (the latter appearing in many AliceBob examples)

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+#### In-flight Transaction
+
+A transaction is considered to be “in-flight” if it has been broadcast but has not yet been included in the Plasma chain.
+A transaction may be in-flight from the perspective of an individual user if that user does not have access to the block in which the transaction is included. 
```
Is it correct to simplify this to:

> A transaction is considered "in-flight" if an individual user does not have access to the block in which the transaction is included.

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+#### Priority
+
+The above game correctly selects the inputs or outputs that are eligible to exit.
+However, we still need to enforce an ordering on exits to ensure that all outputs created by valid transactions will be paid out before any output created by an invalid transaction.
```
This might be a good point to re-iterate that invalid transactions can be exitable.

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+The implementation is designed to be deployed to Ethereum and, as a result, some details of this specification take into account limitations of the EVM.
+
+Additionally, the MoreVP exit protocol is not necessary in all cases.
+Previous work shows that we can use the Plasma MVP exit protocol without confirmation signatures for any transaction included before an invalid (or, in the case of withheld blocks, potentially invalid) transaction.
```
Is there a "previous work" citation we can reference here? Or are you referring to the MVP protocol itself?

smartcontracts replied:
I'm kind of handwaving this here, using MVP w/o confs definitely works in the non-byzantine case but I don't think anyone has formally shown it.

kasima replied:
Maybe we can be more explicitly hand-wavy by using "It can be shown that..."

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+The MoreVP exit protocol described above guarantees that correctly behaving users will always be able to withdraw any funds they hold on the Plasma chain.
+However, we avoided describing how the users actually prove the statements they’re required to prove.
+This section presents a specification for an implementation of the exit protocol.
+The implementation is designed to be deployed to Ethereum and, as a result, some details of this specification take into account limitations of the EVM.
```
Same concerns as @pdobacz about the use of "implementation." I imagine it will easily be confused with a code implementation.

It would be sufficient to use "specification". So the previous two lines could read:

> This section presents a specification for the exit protocol.
> The specification is designed to be deployed to Ethereum and, as a result, some details of this specification take into account limitations of the EVM.

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+The MoreVP exit protocol requires the use of a “challenge-response” mechanism, whereby users can submit a challenge but are subject to a response that invalidates the challenge.
+To give users enough time to respond to a challenge, the exit process is split into two “periods.” When challenges are subject to a response, we require that the challenges be submitted before the end of the first exit period and that responses be submitted before the end of the second.
+We define each period to have a length of half the minimum finalization period (MFP).
+Currently, the MFP is set to 7 days, so each period has a length of 3.5 days.
```
In the Plasma MVP spec, MFP was arbitrarily set to 7 days, so in this case each period has a length of 3.5 days.

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+
+1. Alice spends `UTXO1` in `TX1` to Mallory, creating `UTXO2`.
+2. `TX1` is included in block `N`.
+3. Mallory spends `UTXO2` in `TX2`.
```
Perhaps mention that `TX2` may or may not be in-flight.

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+2. `TX1` is included in block `N`.
+3. Mallory spends `UTXO1` in `TX2`.
+4. `TX2` is included in block `N+M`
+5. Mallory starts an exit referencing `TX1` and places `exit bond`.
```
Why would Mallory do this? And wouldn't Bob want to piggyback?

smartcontracts replied:
Mallory is trying to make the transaction non-canonical so that she can exit the inputs, basically. Assuming the chain is functioning normally, Bob would not want to piggyback (unless Bob wants to exit).

kasima replied:
Ah I see. If that's the case, in step 6, would that "someone" be Mallory, with the chance that there would be no response?

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+7. Bob starts a *standard* exit for `UTXO2`.
+8. Operator’s exit will have priority of position of `UTXO3`.
+Bob’s exit will have priority of position of `UTXO2`.
+9. Bob receives the value of `UTXO2` first, Operator receives the value of `UTXO4` second.
```
This and the next scenario may be misleading. Perhaps add something like: 

> But ideally, all funds have already been exited from the contract by point it's time to process `UTXO4`.

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+1. Mallory spends `UTXO1` in `TX1` to Bob, creating `UTXO2`.
+2. `TX1` is in-flight.
+3. Operator begins withholding blocks while `TX1` is still in-flight.
+4. Bob starts an exit referencing `TX1` and places `exit bond`.
```
This is the same situation as ☝️, right?

> I think it is not rational for Bob to do that, he gains nothing just loses exit_bond. Bob should not process the incoming transaction until he sees it included in the chain.

jepidoptera replied:
Right.  I think Bob should not consider the funds received, and should not attempt an exit, until he sees it confirmed on-chain.  Whatever he's giving Mallory for this money, he should hold onto it until then.

kasima commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+###### Bond Sharing
+
+One way to partially mitigate this attack is for each user who piggybacks to cover some portion of `exit bond`.
+This cuts the per-person value of `exit bond` proportionally to the number of users who have piggybacked.
```
This wouldn't help the Bob in the example though.

smartcontracts replied:
Would help Bob if other honest users piggybacked alongside Bob.

kasima replied:
Right but it's presenting a solution that doesn't necessarily address the problem as stated above. Could cause some confusion. Easy thing would be to add something like, "While it wouldn't mitigate the single output example above, a method for multiple input/output transaction would be..."

smartcontracts committed:
Small changes to morevp.md

kasima commented:
Most of the comments left are nits. The doc feels pretty good where it is.

@kfichter – is there any more you'd like to do?

@pdobacz – how does it look to you?

jepidoptera commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+6. Alice is honest, so she hasn’t spent `UTXO1` in any transaction other than `TX1`.
+7. After period 2, Bob receives the value of `UTXO2`, Alice does not receive the value of `UTXO1`.
+All bonds are refunded.
+
```
Why did Alice place a bond? She was never going to get anything back.

smartcontracts replied:
Yeah Alice doesn't need to place a bond here. She would want to place a bond if there was another untrusted input.

smartcontracts replied:
I'll change that.

jepidoptera commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,655 @@
+If any transactions have been presented during the first period, any other user can respond to the challenge by proving that the exiting transaction is actually included in the chain before the oldest presented competing transaction.
+If this response is given before the end of the second period, then the exiting transaction is determined to be canonical and the responder receives the `exit bond` placed by the user who started the exit.
+Otherwise, the exiting transaction is determined to be non-canonical and the challenger receives `exit bond`.
+
```
This doesn't seem right. The person who started the exit loses their bond when someone else proves that their exit _is_ canonical?

smartcontracts replied:
This is correct. Person loses the bond because other person has to expend gas to respond to the challenge.

jepidoptera replied:
But that never happens unless the exitor has double-spent an input, so... ok.

smartcontracts committed:
Small edits to morevp.md

smartcontracts committed:
Add link to original morevp.md

jepidoptera commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,662 @@
+Owners of outputs on the Plasma chain should be able to start an exit via either mechanism, but not both.
+Although the two protocols use different determinations for exit priority, we still need a total ordering on exits.
+Therefore, every exit, no matter the protocol used, must be included in the same priority queue for processing.
+
```
Is there any scenario where using youngest input as the priority criterion for a "standard" exit creates a problem?

pdobacz replied:
If you'd do that, i.e. have youngest input priority but without the in-flight exit protocol (exit game), the operator would be able to multi-spend an old output and exit all of the multi spending transactions, with higher priority than "honest exits".

With the more complex/costly MoreVP exit protocol, there is no such possibility.

jepidoptera replied:
But let's say we leave the in-flight exit protocol as-is.  The operator will be challenged for producing multiple competing exit transactions, and all will be rolled back at operator's expense.

I'm just looking for a way to reconcile the ordering system.

I'm trying to grasp the fundamental difference between the "standard" and "moreVP" exit protocols, and whether it's true that the former requires a different priority ordering (tx # vs. youngest input) to work properly.  If not, then combining them into the same queue becomes trivial.

smartcontracts replied:
They use a different priority ordering but they can still be combined into the same queue.

pdobacz replied:
@jepidoptera I think the sequence of implications is for

> "standard" and "moreVP" exit protocols, and whether it's true that the former requires a different priority ordering (tx # vs. youngest input)

is this: moreVP protocol is expensive => we keep the standard exit => standard exit uses output position for ordering => we don't care about inputs to that transaction having been valid or not

or the other way around: in-flight transaction's outputs may have potentially low priority => in-flight transaction's outputs are doomed in standard exit proto => need to use their inputs as "source of higher priority" => need to prove more => more expensive, moreVP protocol for in-flight txs.

This is how I got the hang of that difference, does this help?

pdobacz commented:
@kasima @kfichter there are still some comments that I'd like to look into anddiscuss. GH is giving me a hard time referencing them (outdated), so I'll just drop them here again:

1/ around section "Priority":

```
I also think detailed interactions between the two exit protocols need being specified, namely that you
cannot in-flight exit an output that is being exited regularly, probably vice versa + other more exotic
and gory scenarios

contd:
what about the above paragraph, do you think we should be more specific about the correctness of
these interactions? is there a list of conditions that must be checked extra for the two protocols to
not interfere badly?
```

2/ 
```
@kfichter note that there are two sections called "Exit Protocol" now. So if you agree with my
argumentation to not use the word "implementation" in this context, the initial section should
be renamed
```

3/ 
```
Id define MEP/2 as a new abbreviation (same suggestion as with limbo cancels) and state explicitly,
that with this protocol in place we require, that all watchers validate the chain at least every,
not every MEP as previously required

contd:
Oops, sorry, now I see my comment got severed somehow. It was supposed to say "watchers validate
the chain at least every min(REP, /new abbreviation/), not every min(MFP, REP) as previously required" -
with some extra corrections applied
```

smartcontracts committed:
Update morevp.md

smartcontracts committed:
Edit morevp.md

smartcontracts committed:
Minor edits to MoreVP.md

smartcontracts committed:
Update morevp.md

Slight update to handle the case where users try to exit the same inputs in multiple txs.

josojo commented on plasma/plasma-mvp/specifications/morevp.md:
```diff
@@ -0,0 +1,676 @@
+2. `TX1` is in-flight.
+3. Operator begins withholding blocks while `TX1` is still in-flight.
+Neither Alice nor Bob know if the transaction has been included in a block.
+4. Someone with access to `TX1` (Alice, Bob, or otherwise) starts an exit referencing `TX1` and places `exit bond`.
```
Maybe it would make sense to mention that Alice should start the exit. Yes, everyone can do it, but only Alice knows whether the transaction is canonical or not. Hence, she is the only one, who knows that the exit bond will not be slashed in any case.
Does this makes sense, or did I misunderstand something? @kfichter
